pkg/operator/customroute/custom_route_controller.go

```diff
@@ -0,0 +1,283 @@
+	if componentRoute := routecomponenthelpers.GetComponentRouteSpec(ingressConfig, route.ObjectMeta.Namespace, route.ObjectMeta.Name); componentRoute != nil {
+		var errors []error
+		// Check if the provided secret is valid
+		secretName = componentRoute.ServingCertKeyPairSecret.Name
+		if err := c.validateCustomTLSSecret(secretName); err != nil {
+			errors = append(errors, err)
+		}
+
+		// Check if the provided hostname is valid
+		hostname := string(componentRoute.Hostname)
+		if _, err := url.Parse(hostname); err != nil {
+			errors = append(errors, err)
+		}
+
+		if errors != nil {
+			return nil, "", errors
+		}
+
+		route.Spec.Host = hostname
+	}
+
+	return route, secretName, nil
+}
+
+func (c *customRouteController) validateCustomTLSSecret(secretName string) error {
+	if secretName != "" {
+		secret, err := c.secretLister.Secrets("openshift-config").Get(secretName)
+		if err != nil {
+			return err
+		}
+
+		var errors []error
+		privateKeyData, ok := secret.Data[corev1.TLSPrivateKeyKey]
+		if !ok {
+			errors = append(errors, fmt.Errorf("custom route secret must include key %s", corev1.TLSPrivateKeyKey))
+		} else {
+			errors = append(errors, certs.ValidatePrivateKey(privateKeyData)...)
+		}
+
+		certData, ok := secret.Data[corev1.TLSCertKey]
+		if !ok {
+			errors = append(errors, fmt.Errorf("custom route secret must include key %s", corev1.TLSCertKey))
+		} else {
+			errors = append(errors, certs.ValidateServerCert(certData)...)
+		}
+
+		if len(errors) != 0 {
+			return fmt.Errorf("error validating secret %s/%s: %v", "openshift-config", secretName, errors)
+		}
+	}
+	return nil
+}
+
+func (c *customRouteController) updateIngressConfigStatus(ctx context.Context, ingressConfig *configv1.Ingress, customRouteErrors []error) error {
+	// update ingressConfig status
+	// TODO: consider using c.targetRoute. Nevertheless the downside is that we might have a stale reference?
+	route, err := c.routeLister.Routes(c.targetRoute.ObjectMeta.Namespace).Get(c.targetRoute.ObjectMeta.Name)
+	if err != nil {
+		return err
+	}
+
+	componentRoute := &configv1.ComponentRouteStatus{
+		Namespace:        c.componentRoute.Namespace,
+		Name:             c.componentRoute.Name,
+		DefaultHostname:  configv1.Hostname(c.targetRoute.ObjectMeta.Name + "." + ingressConfig.Spec.Domain),
+		CurrentHostnames: []configv1.Hostname{configv1.Hostname(route.Spec.Host)},
+		ConsumingUsers:   c.consumingUsers,
+		RelatedObjects: []configv1.ObjectReference{
+			{
+				Group:     routev1.GroupName,
+				Resource:  "routes",
+				Name:      c.targetRoute.ObjectMeta.Name,
+				Namespace: c.targetRoute.ObjectMeta.Namespace,
+			},
+		},
+	}
+	conditions := checkErrorsConfiguringCustomRoute(customRouteErrors)
+	if conditions == nil {
+		conditions = checkIngressURI(ingressConfig, route)
+		if conditions == nil {
+			conditions = checkRouteAvailablity(c.secretLister, ingressConfig, route)
+		}
+	}
+	componentRoute.Conditions = ensureDefaultConditions(conditions)
```

Comment:
In our experience, this construction is chronically buggy.  The sync method where you do work shouldn't set the status because when it fails (and it will fail), it doesn't keep the status up to date.  If you want to wrap your main sync method in something else *and* be sure to handle panics in it, you can try to set both conditions.

It seems slightly more reasonable to update the controller factory bit to always handle this.

Reply:
This makes a lot of sense. Let me divide the code in the `sync` method into 2 sections. One of them will be creating the Route and requesting the Secret sync and the other will handle the status updates. 

It's worth to mention that this approach minimizes the chances of getting Conditions out of sync. Nevertheless, if a bug happens in setting the Conditions, you will never know that (unless you analyze the logs).

Reply:
During the refactoring I also changed the semantics a bit. I will need to verify the integration tests once we're happy with this PR.